server/etcdserver/api/etcdhttp/readiness.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdhttp
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"go.uber.org/zap"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/server/v3/auth"
+)
+
+func InstallHealthEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	InstallLivezEndpoints(lg, mux, server)
+	InstallReayzEndpoints(lg, mux, server)
+}
+
+type HealthCheck func(ctx context.Context) error
+
+type CheckRegistry struct {
+	path string
+
+	checks map[string]HealthCheck
+	// the list of checkNames are needed to ensure the order of checks.
+	checkNames []string
+}
+
+func InstallLivezEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	reg := CheckRegistry{path: "/livez", checks: make(map[string]HealthCheck)}
+	reg.Register("serializable_read", checkSerializableRead(server))
+	reg.InstallHttpEndpoints(lg, mux)
+}
+
+func InstallReayzEndpoints(lg *zap.Logger, mux *http.ServeMux, server ServerHealth) {
+	reg := CheckRegistry{path: "/readyz", checks: make(map[string]HealthCheck)}
+	reg.Register("data_corruption", checkAlarm(server, etcdserverpb.AlarmType_CORRUPT))
+	reg.Register("serializable_read", checkSerializableRead(server))
+	reg.InstallHttpEndpoints(lg, mux)
+}
+
+func (reg *CheckRegistry) Register(name string, check HealthCheck) {
+	reg.checkNames = append(reg.checkNames, name)
+	reg.checks[name] = check
+}
+
+func (reg *CheckRegistry) InstallHttpEndpoints(lg *zap.Logger, mux *http.ServeMux) {
+	lg.Info(fmt.Sprintf("Installing %s http endpoint.", reg.path))
+	mux.Handle(reg.path, reg.handleRootHealth(lg))
+	for checkName, check := range reg.checks {
+		mux.Handle(fmt.Sprintf("%s/%v", reg.path, checkName), adaptCheckToHandler(check))
```

Comment:
`adaptCheckToHandler` function feels a little akward. Have you thought on instead having a `InstallCheck` method on `CheckRegistry`. It could look like 

```
func (reg *CheckRegistry) InstallHttpEndpoints(lg *zap.Logger, mux *http.ServeMux) {
	lg.Info(fmt.Sprintf("Installing %s http endpoint.", reg.path))
   mux.Handle(reg.path, reg.handleRootHealth(lg))
	for checkName, check := range reg.checks {
	    reg.installCheckEndpoint(checkName, check, mux)
	}
}

func (reg *CheckRegistry) installCheckEndpoint(name string, check HealthCheck, mux *http.ServeMux) {
  mux.Handle(path.Join(reg.path, name), func(w http.ResponseWriter, r *http.Request) {
		err := check(r.Context())
		if err != nil {
			http.Error(w, fmt.Sprintf("internal server error: %v", err), http.StatusInternalServerError)
		} else {
			fmt.Fprint(w, "ok")
		}
	})
}
```

Reply:
done